File: tasks/native-runtime/store_parity_probe.py
"""Store parity probe: the budget database and the file index.

Unlike the earlier probes, this one exercises **real I/O on both sides** — each side creates
its own SQLite database and its own cache files in a temporary directory, and the outputs are
compared byte for byte. That makes three things comparable that a stub could not:

- the DDL, because SQLite stores the statement text verbatim in `sqlite_master`;
- the accumulated row, because the upsert is what makes two calls add up;
- the four failure modes of `load_cached_file`, message and code included.

`updated_at` is deliberately left out of the raw-row view: the oracle reads the wall clock
there and its value is not part of the contract the two sides can share.

Usage::

    python tasks/native-runtime/store_parity_probe.py > python.json
    cd rust && cargo run -p deepseek-policy --example store_parity_probe > ../rust.json
"""

from __future__ import annotations

import json
import sqlite3
import sys
import tempfile
from pathlib import Path
from typing import Any, Callable

REPO = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(REPO))

from deepseek_infra.core.errors import AppError  # noqa: E402
from deepseek_infra.infra.gateway import budget_manager as bm  # noqa: E402
from deepseek_infra.infra.rag import files  # noqa: E402

BAD_FILE_IDS = ["A" * 32, "a" * 31, "a" * 33, "g" * 32, "", "a" * 31 + "B"]

RAW_COLUMNS = [
    "scope",
    "day",
    "prompt_tokens",
    "completion_tokens",
    "cost_usd",
    "model_calls",
    "search_calls",
    "tool_calls",
]


def error_view(call: Callable[[], Any]) -> Any:
    try:
        return {"ok": call()}
    except AppError as exc:
        return {"error": str(exc), "code": exc.code.value, "status": exc.status}


def main() -> int:
    out: dict[str, Any] = {}
    temp_dir = Path(tempfile.mkdtemp(prefix="store-probe-"))

    # --- the budget database -----------------------------------------------------------
    saved = (bm.BUDGET_DIR, bm.BUDGET_DB, bm.BUDGET_TRACKING_ENABLED, files.FILE_CACHE_DIR, files.PROJECTS_DIR)
    try:
        bm.BUDGET_DIR = temp_dir / "budget"
        bm.BUDGET_DB = bm.BUDGET_DIR / "budget.db"
        bm.BUDGET_TRACKING_ENABLED = True

        connection = bm.connect_db()
        bm.initialize_schema(connection)
        schema_row = connection.execute(
            "SELECT sql FROM sqlite_master WHERE type = 'table' AND name = ?", (bm.SPEND_TABLE,)
        ).fetchone()
        out["budget::schema"] = schema_row[0]
        connection.close()
        out["budget::dir-created"] = bm.BUDGET_DIR.is_dir()

        bm.record_spend("global", prompt_tokens=10, completion_tokens=20, cost_usd=4.9e-05, model_calls=1)
        bm.record_spend("global", prompt_tokens=5, cost_usd=1e-06, model_calls=1, tool_calls=2)
        out["budget::accumulated"] = bm.daily_spend("global")
        out["budget::other-scope"] = bm.daily_spend("project-x")
        out["budget::other-day"] = bm.daily_spend("global", "2026-01-01")

        connection = bm.connect_db()
        raw = connection.execute(
            "SELECT * FROM budget_daily WHERE scope = ? AND day = ?", ("global", bm.today())
        ).fetchone()
        out["budget::raw-row"] = {name: raw[name] for name in RAW_COLUMNS}
        connection.close()

        # A fresh database directory is created on demand, so a first write to a new day and
        # scope lands in a row of its own.
        bm.record_spend("project-x", prompt_tokens=1, model_calls=1)
        out["budget::new-scope"] = bm.daily_spend("project-x")

        # --- the file index ------------------------------------------------------------
        files.FILE_CACHE_DIR = temp_dir / "cache"
        files.PROJECTS_DIR = temp_dir / "projects"
        files.FILE_CACHE_DIR.mkdir(parents=True, exist_ok=True)

        file_id = "a" * 32
        document = {"id": file_id, "name": "报告.pdf", "chunks": [{"text": "片段"}]}
        (files.FILE_CACHE_DIR / f"{file_id}.json").write_text(
            json.dumps(document, ensure_ascii=False), encoding="utf-8"
        )
        out["file::ok"] = files.load_cached_file(file_id)

        for index, bad in enumerate(BAD_FILE_IDS):
            out[f"file::bad-id-{index}"] = error_view(lambda bad=bad: files.load_cached_file(bad))

        out["file::missing"] = error_view(lambda: files.load_cached_file("b" * 32))

        (files.FILE_CACHE_DIR / f"{'c' * 32}.json").write_text("not json", encoding="utf-8")
        out["file::not-json"] = error_view(lambda: files.load_cached_file("c" * 32))

        (files.FILE_CACHE_DIR / f"{'d' * 32}.json").write_text("[1, 2]", encoding="utf-8")
        out["file::not-object"] = error_view(lambda: files.load_cached_file("d" * 32))

        # A project cache lives under its own directory, and the project id is validated
        # before the path is built.
        project_files = files.PROJECTS_DIR / "proj1" / "files"
        project_files.mkdir(parents=True, exist_ok=True)
        (project_files / f"{file_id}.json").write_text(
            json.dumps({**document, "projectId": "proj1"}, ensure_ascii=False), encoding="utf-8"
        )
        out["file::project-ok"] = files.load_cached_file(file_id, "proj1")
        out["file::project-short"] = error_view(lambda: files.load_cached_file(file_id, "ab"))
        out["file::project-empty-is-global"] = files.load_cached_file(file_id, "")

        # The document as it stands now, after a rewrite: a changed file changes the memo key.
        (files.FILE_CACHE_DIR / f"{file_id}.json").write_text(
            json.dumps({**document, "name": "报告-v2.pdf"}, ensure_ascii=False), encoding="utf-8"
        )
        out["file::after-rewrite"] = files.load_cached_file(file_id)
    finally:
        (
            bm.BUDGET_DIR,
            bm.BUDGET_DB,
            bm.BUDGET_TRACKING_ENABLED,
            files.FILE_CACHE_DIR,
            files.PROJECTS_DIR,
        ) = saved

    json.dump(out, sys.stdout, ensure_ascii=False, indent=2, sort_keys=True)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: rust/crates/deepseek-policy/examples/store_parity_probe.rs
//! Store parity probe, Rust side: a real SQLite database and a real file cache, like the
//! Python side, compared byte for byte.
//!
//! `updated_at` is left out of the raw-row view: the oracle reads the wall clock there, so its
//! value cannot match and is not contract.
//!
//! Usage::
//!
//!     python tasks/native-runtime/store_parity_probe.py > python.json
//!     cd rust && cargo run -p deepseek-policy --example store_parity_probe > ../rust.json

use std::path::PathBuf;

use rusqlite::Connection;
use serde_json::{Map, Value, json};

use deepseek_policy::app_error::AppError;
use deepseek_policy::budget_ledger::{LedgerDeps, daily_spend, record_spend};
use deepseek_policy::budget_manager::BudgetSettings;
use deepseek_policy::budget_store::{BudgetStore, connect_db, initialize_schema};
use deepseek_policy::file_store::FileStore;
use deepseek_policy::python_json::OrderedJson;

const DAY: &str = "2026-09-18";
const NOW_ISO: &str = "2026-09-18T00:00:00Z";

const RAW_COLUMNS: [&str; 8] = [
    "scope",
    "day",
    "prompt_tokens",
    "completion_tokens",
    "cost_usd",
    "model_calls",
    "search_calls",
    "tool_calls",
];

fn error_view(result: Result<Value, AppError>) -> Value {
    match result {
        Ok(value) => json!({"ok": value}),
        Err(error) => json!({"error": error.message, "code": error.code, "status": error.status}),
    }
}

/// The success case renders the document itself, as the oracle returns it; only the failures
/// get the code/status envelope.
fn ok_or_error(result: Result<Value, AppError>) -> Value {
    match result {
        Ok(value) => value,
        Err(error) => json!({"error": error.message, "code": error.code, "status": error.status}),
    }
}

fn bad_file_ids() -> Vec<String> {
    vec![
        "A".repeat(32),
        "a".repeat(31),
        "a".repeat(33),
        "g".repeat(32),
        String::new(),
        format!("{}B", "a".repeat(31)),
    ]
}

fn temp_dir(tag: &str) -> PathBuf {
    let base = std::env::temp_dir().join(format!("store-probe-{tag}-{}", std::process::id()));
    let _ = std::fs::remove_dir_all(&base);
    std::fs::create_dir_all(&base).expect("a temp directory");
    base
}

fn raw_row(connection: &Connection, scope: &str, day: &str) -> Value {
    let mut statement = connection
        .prepare("SELECT * FROM budget_daily WHERE scope = ? AND day = ?")
        .expect("the select prepares");
    let mut rows = statement
        .query(rusqlite::params![scope, day])
        .expect("the select runs");
    let Some(row) = rows.next().expect("a row result") else {
        return Value::Null;
    };
    let mut out = Map::new();
    for name in RAW_COLUMNS {
        let index = row
            .as_ref()
            .column_names()
            .iter()
            .position(|column| *column == name)
            .expect("the column exists");
        let value = match row.get_ref(index).expect("the column reads") {
            rusqlite::types::ValueRef::Null => Value::Null,
            rusqlite::types::ValueRef::Integer(int) => Value::from(int),
            rusqlite::types::ValueRef::Real(float) => json!(float),
            rusqlite::types::ValueRef::Text(bytes) => {
                Value::String(String::from_utf8_lossy(bytes).into_owned())
            }
            rusqlite::types::ValueRef::Blob(bytes) => {
                Value::Array(bytes.iter().map(|byte| Value::from(*byte)).collect())
            }
        };
        out.insert(name.to_string(), value);
    }
    Value::Object(out)
}

fn main() {
    let mut out: Map<String, Value> = Map::new();
    let temp = temp_dir("all");

    // --- the budget database -----------------------------------------------------------
    let budget_dir = temp.join("budget");
    let budget_db = budget_dir.join("budget.db");
    let store = BudgetStore::new(&budget_dir, &budget_db);

    out.insert(
        "budget::schema".to_string(),
        json!(
            store
                .schema_sql()
                .expect("the schema reads")
                .unwrap_or_default()
        ),
    );
    out.insert(
        "budget::dir-created".to_string(),
        json!(budget_dir.is_dir()),
    );

    let read = |scope: &str, day: &str| store.read_spend_row(scope, day);
    let write = |row: &Value| store.write_spend_row(row);
    let settings = BudgetSettings::default();
    let writing = LedgerDeps {
        database_present: false,
        database_path: store.database_path(),
        day: DAY.to_string(),
        now_iso: NOW_ISO.to_string(),
        read_spend_row: &read,
        write_spend_row: &write,
    };
    assert!(
        record_spend(
            &json!("global"),
            &json!({"prompt_tokens": 10, "completion_tokens": 20, "cost_usd": 4.9e-05, "model_calls": 1}),
            &settings,
            &writing,
        )
        .is_none()
    );
    assert!(
        record_spend(
            &json!("global"),
            &json!({"prompt_tokens": 5, "cost_usd": 1e-06, "model_calls": 1, "tool_calls": 2}),
            &settings,
            &writing,
        )
        .is_none()
    );
    let reading = LedgerDeps {
        database_present: store.database_present(),
        database_path: store.database_path(),
        day: DAY.to_string(),
        now_iso: NOW_ISO.to_string(),
        read_spend_row: &read,
        write_spend_row: &write,
    };
    let (accumulated, _) = daily_spend("global", None, &reading);
    out.insert("budget::accumulated".to_string(), accumulated);
    let (other_scope, _) = daily_spend("project-x", None, &reading);
    out.insert("budget::other-scope".to_string(), other_scope);
    let (other_day, _) = daily_spend("global", Some("2026-01-01"), &reading);
    out.insert("budget::other-day".to_string(), other_day);

    let connection = connect_db(&budget_dir, &budget_db).expect("the database opens");
    initialize_schema(&connection).expect("the schema applies");
    out.insert(
        "budget::raw-row".to_string(),
        raw_row(&connection, "global", DAY),
    );
    drop(connection);

    assert!(
        record_spend(
            &json!("project-x"),
            &json!({"prompt_tokens": 1, "model_calls": 1}),
            &settings,
            &writing,
        )
        .is_none()
    );
    let (new_scope, _) = daily_spend("project-x", None, &reading);
    out.insert("budget::new-scope".to_string(), new_scope);

    // --- the file index ----------------------------------------------------------------
    let cache_dir = temp.join("cache");
    let projects_dir = temp.join("projects");
    std::fs::create_dir_all(&cache_dir).expect("the cache directory");
    let files = FileStore::new(&cache_dir, &projects_dir);

    let file_id = "a".repeat(32);
    let document = json!({"id": file_id, "name": "报告.pdf", "chunks": [{"text": "片段"}]});
    std::fs::write(
        cache_dir.join(format!("{file_id}.json")),
        serde_json::to_string(&document).expect("the document serialises"),
    )
    .expect("the document writes");
    out.insert(
        "file::ok".to_string(),
        ok_or_error(files.load_cached_file(&file_id, None)),
    );

    for (index, bad) in bad_file_ids().iter().enumerate() {
        out.insert(
            format!("file::bad-id-{index}"),
            error_view(files.load_cached_file(bad, None)),
        );
    }
    out.insert(
        "file::missing".to_string(),
        error_view(files.load_cached_file(&"b".repeat(32), None)),
    );

    std::fs::write(
        cache_dir.join(format!("{}.json", "c".repeat(32))),
        "not json",
    )
    .expect("the malformed document writes");
    out.insert(
        "file::not-json".to_string(),
        error_view(files.load_cached_file(&"c".repeat(32), None)),
    );
    std::fs::write(cache_dir.join(format!("{}.json", "d".repeat(32))), "[1, 2]")
        .expect("the array document writes");
    out.insert(
        "file::not-object".to_string(),
        error_view(files.load_cached_file(&"d".repeat(32), None)),
    );

    let project_files = projects_dir.join("proj1").join("files");
    std::fs::create_dir_all(&project_files).expect("the project directory");
    let mut projected = document.as_object().cloned().unwrap_or_default();
    projected.insert("projectId".to_string(), json!("proj1"));
    std::fs::write(
        project_files.join(format!("{file_id}.json")),
        serde_json::to_string(&Value::Object(projected)).expect("the document serialises"),
    )
    .expect("the project document writes");
    out.insert(
        "file::project-ok".to_string(),
        ok_or_error(files.load_cached_file(&file_id, Some("proj1"))),
    );
    out.insert(
        "file::project-short".to_string(),
        error_view(files.load_cached_file(&file_id, Some("ab"))),
    );
    out.insert(
        "file::project-empty-is-global".to_string(),
        ok_or_error(files.load_cached_file(&file_id, Some(""))),
    );

    let mut rewritten = document.as_object().cloned().unwrap_or_default();
    rewritten.insert("name".to_string(), json!("报告-v2.pdf"));
    std::fs::write(
        cache_dir.join(format!("{file_id}.json")),
        serde_json::to_string(&Value::Object(rewritten)).expect("the document serialises"),
    )
    .expect("the rewrite lands");
    out.insert(
        "file::after-rewrite".to_string(),
        ok_or_error(files.load_cached_file(&file_id, None)),
    );

    let rendered = OrderedJson::from_value_with_order(&Value::Object(out), &[]).render_indent_2();
    println!("{rendered}");
}
